python/cudf/cudf/core/frame.py

```diff
@@ -114,6 +115,19 @@ def _index_names(self) -> List[Any]:  # TODO: List[str]?
     def _columns(self) -> List[Any]:  # TODO: List[Column]?
         return self._data.columns
 
+    def _column_name_to_indices(self, column_names, include_index=True):
+        column_names = set(column_names)
+        idx_len = (
+            len(self._index._data)
+            if include_index and self._index is not None
+            else 0
+        )
+        return [
+            i + idx_len
+            for i, name in enumerate(self._data.keys())
+            if name in column_names
+        ]
```

Comment:
Python's `set` is actually not insertion ordered. While dictionaries have been guaranteed to be insertion-ordered as of 3.7+, sets are not ordered and there is AFAIK no plan to change that. The hash table implementations used are quite different and optimized for different use cases.

We _do_ implicitly rely on the ordering guarantee of dictionaries. That was a conscious decision made for performance reasons in #7847. The OrderedDict class is substantially slower than the dict, and given that the majority of our algorithms involve operations on the dictionary of columns underlying a frame we see pretty significant performance improvements from the switch.

However, it's worth noting that in fact Michael's original code _does_ preserve the original order and does not depend on any ordering of the set. You'll notice that his original code was iterating over `self._data` (a dict (well really a ColumnAccessor that stores a dict under the hood), which is insertion ordered) and simply checking set membership in `column_names`. Your suggestion, however, does rely on ordering because it is iterating over the set of column names. That change would indeed be unsafe and break ordering guarantees. Typically the best approach for something like this where you need an ordered set is to instead use something like `dict.fromkeys(column_names)`, which will generate a (insertion-ordered) dictionary with all values set to `None` and can therefore substitute for the lack of order in sets.

Reply:
Sorry, my previous comment was wrong. Thanks @vyasr for setting it straight. I would add a comment to indicate that the returned order of indices is that of the `Frame`, not the input order of `column_names`.

```suggestion
    def _column_name_to_indices(self, column_names, include_index=True):
        # Return indices of the provided column names, offset by the number of
        # index columns if include_index is True. The order of indices returned
        # corresponds to the column order in this Frame.
        column_names = set(column_names)
        idx_len = (
            len(self._index._data)
            if include_index and self._index is not None
            else 0
        )
        return [
            i + idx_len
            for i, name in enumerate(self._data.keys())
            if name in column_names
        ]
```